src/Analyser/TypeSpecifier.php

```diff
@@ -1363,8 +1363,16 @@ private function createForExpr(
 		$exprString = $this->exprPrinter->printExpr($expr);
 		if ($context->false()) {
 			$sureNotTypes[$exprString] = [$expr, $type];
+			if ($expr !== $originalExpr) {
```

Comment:
I would rather compare the string representations of those expressions.

Reply:
Adapted. On my phone 🙃